### DISCUSSION TITLE
Style: style prop could directly accept array

### DISCUSSION THREAD

**Discussion by MoOx** [BuckleTypes/bs-react-native#750] [Ideas]
At first, when we revisited bs-react-native, we had for style
- single style
- array of style
- list of style

Recently, since array usage has been promoted in favour of list (we made this a breaking change in one of the last major release), so we now only have single style or array of style.

**I find myself using array of style more than half of the time** (because I use a combination of static styles and dynamic styles (hello dark mode).
I would be interested to know about the community practise because I am thinking

 _**"what if style prop only accept array of style?"**_

Currently we need to write:
- single:  `<Text style={style}`
- multiple:  `<Text style={array([style, style1])}` or `<Text style={Style.array([style, style1])}` (depending on opened modules)

If we change this, we would just have to type
- single: `<Text style={[style]}`
- multiple: `<Text style={[style, style2]}`

That would be obviously a breaking change, but should be easy to change.

Thoughts?  poke @jfrolich @cknitt @sgny @idkjs @Freddy03h @gedeagas @celsobonutti (and anyone who wants to give their feedback on this idea)

**Comment by jfrolich** [comment]
Same feeling @MoOx, also since using reanimated2 I use array even more. I think this would be a net improvement. There might be a small performance regression when using a an array of 1 vs a single style, but probably diminishable.

**Comment by Freddy03h** [comment]
Same for me, I'll not be against this. But what about `arrayOption` ?

**Reply by jfrolich** [reply]
`keepMap` but also it should keep working because react-native accepts nested arrays for styles

**Reply by MoOx** [reply]
Actually Style.t is used for everything, so a `viewStyle` is a `t`, and `array(...)` is t too, same for `arrayOption`. This means if we change style props from `Style.t` to `array(Style.t)`, you could still use `array` and `arrayOption`.
Like said above, this is due to the fact that RN accepts nested arrays, so you can pass in JS `style={[[{color: "blue"}], [[{"backgroundColor": "red"}]]]}` and it will work.

**Reply by sgny** [reply]
Exactly. Otherwise we would have already run into trouble seeing how helper functions allow recasting `array(Style.t)` into `Style.t`.

I like this change, but I further propose defining `Style.style` to  return `array<t>` instead of `t` as below. Then, minimal refactoring would needed and we can still continue writing `style=someStyle`. The resulting type definitions would not be less precise in practice either.

```
style: (
  ~resizeMode: resizeMode=?,
  ...
  ~zIndex: int=?,
  unit,
) => array<t> = ""
```

**Reply by MoOx** [reply]
Oh that is actually an interesting idea but maybe that would allow people to write things that will create runtime error like `someStyle->Array.length` ?!

**Reply by sgny** [reply]
That's a valid concern. Multiple styles would become a worse problem with this too. Joining an actual array and an element would be problematic.

I suppose I haven't thought this through.

**Reply by jfrolich** [reply]
Yes this would not be safe

**Comment by gedeagas** [comment]
I also would not be against this. Should be an easy refactor on the existing codebase.

**Comment by MoOx** [comment]
Since time is limited, I have a simple solution (temporary) for us to try this at a large scale before pushing this as the main method: I was thinking we could add an alternate Style module (eg: Style2 or AStyle or a better (but short) name) to try this. Obviously if we keep Style that would means that there is at some point a weird magic trick to make them coexist but maybe it worth a try? Any thoughts?

**Comment by cknitt** [comment]
Personally, I strongly prefer the way it is currently implemented.

I just checked one of our projects and, out of 320 usages of `style={`, found just

* 38 x `Style.array`
* 7 x `Style.arrayOption`

Also, usage with `arrayOption` within an array would seem weird to me.

**Reply by Freddy03h** [reply]
I'm still thinking style should accept an array directly

On my project:

944 usages of `style={`

- 395 x `Style.array`
- 64 x `Style.arrayOption`

Like @MoOx said, there is too much dynamic styles on a mobile today: light and dark mode, safe area insets, screen dimensions, ios vs android vs web, …

**Comment by MoOx** [comment]
Interesting stuff are in this PR: we tried but there is some complications: runtime & futur proof issues? 

https://github.com/rescript-react-native/rescript-react-native/pull/769

@Freddy03h you mentioned that RNW could go back to num ids. But they just changed in 0.18 https://github.com/necolas/react-native-web/releases/tag/0.18.0

**Reply by Freddy03h** [reply]
Yeah I just remember I read something about it, but it say 

> accessing the returned style objects at runtime is still not recommended as it can prevent static extraction to CSS

so my point about using spread operator is still irrelevant.
